### ISSUE TITLE
String format TypeError prevents history table from rendering

### ISSUE THREAD

**Issue by jamesachamp** [internetarchive/openlibrary#3883]
When users navigate to https://openlibrary.org/type/work a flash error message appears and the edit history is not rendered.

### Evidence / Screenshot (if possible)
![typeError](https://user-images.githubusercontent.com/28732543/95096440-44281400-06fa-11eb-91c3-45b53aefe47f.png)

### Relevant url?

https://openlibrary.org/type/work

### Steps to Reproduce

1. Navigate to https://openlibrary.org/type/work

* Actual: Error message is displayed and history table is not displayed.
* Expected: History table is displayed.

### Details

- **Logged in (Y/N)?** N/A
- **Browser type/version?** N/A
- **Operating system?** N/A
- **Environment (prod/dev/local)?** prod

### Related files

Error occurs at line 26 of `/openlibrary/templates/lib/history.html`

**Comment by nk4456542** [comment]
Hello, Can you assign me this issue, I am a beginner to Open-source in general. But I think I can solve it with a bit of help.

**Comment by mekarpeles** [comment]
You can see the stack-trace by going to https://openlibrary.org/type/work?debug=true (incognito)

![openlibrary org_type_work_debug=true](https://user-images.githubusercontent.com/978325/95172370-124f9580-076c-11eb-91dc-1f9d9e0c9ee2.png)

**Comment by nk4456542** [comment]
Thanks for the help @mekarpeles, looking into it. Will try to solve and ask for help if needed.

**Comment by nk4456542** [comment]
### Asking for help: If anyone is free, please go through this

### the history section is not being rendered to the following pages too: 'type/collection','type/place', 'type/scan_location', and some more. These are the pages that I could remember at the top of my head. 

The error goes away when we replace %d placeholder with the %s placeholder in the following line:
Original line:
       `<li class="link inline"><a rel="nofollow" href="$url">$ungettext('1 revision', '%(count)d revisions', latest.revision, count=latest.revision)</a></li> `
Modified line:
` <li class="link inline"><a rel="nofollow" href="$url">$ungettext('1 revision', '%(count)s revisions', latest.revision, count=latest.revision)</a></li>`

After replacing the former(%d) line with the latter(%s), history section gets rendered for all the pages ( mentioned above ) where this issue is present and no error is displayed, but this is not the solution, because the number of revisions are not being displayed as I have made the count variable as string:

![Screenshot from 2020-10-06 20-15-10](https://user-images.githubusercontent.com/57295877/95217463-d4587d00-0810-11eb-9a55-a0ce7301b92d.png)

### I am stuck here : 
### The stack-trace is not showing the values of the variables (h,versions,latest) in the 'type/work' page:
I opened this https://openlibrary.org/type/work?debug=true link in the incognito mode itself ( Browser:Chrome, Version 85.0.4183.121 (Official Build) (64-bit), OS = Ubuntu 20.04 )

### tldr; My question: 
### What should I do to find the variable values of the 'type/work' page?
![Screenshot from 2020-10-06 21-38-51](https://user-images.githubusercontent.com/57295877/95227986-63b75d80-081c-11eb-9e36-aafeb13e3a8c.png)

This is the screenshot through the local setup ie, through 0.0.0.0:8080, but the values are not shown even in the development link (https://openlibrary.org/type/work?debug=true). You can see the image that @mekarpeles has commented.


### What I did to solve the issue:
### I looked at the stack-trace as @mekarpeles told me to, but the variables "h, versions and latest" do not show any values when looking at the  https://openlibrary.org/type/work?debug=true, so I did the following steps:
 1. Took a look at the get_history functon
https://github.com/internetarchive/openlibrary/blob/3be8c39ea5352170d1541763ce8e9dbaa32714e6/openlibrary/plugins/upstream/utils.py#L357
2. Also at the get_history_preview function 
https://github.com/internetarchive/openlibrary/blob/3be8c39ea5352170d1541763ce8e9dbaa32714e6/openlibrary/core/models.py#L72
3. And finally at the function that was being called ungettext function
https://github.com/internetarchive/openlibrary/blob/3be8c39ea5352170d1541763ce8e9dbaa32714e6/openlibrary/i18n/__init__.py#L150
4. Then I introduced a little bug to see if the local variables of the '/type' (because history section was being rendered ont the '/type' page ) are being shown in the debug=true mode, the variables are being shown in the '/type' page but not '/type/work' page ( a little weird ), the values of variables of '/type' page:

h = {'lastest_revision': 1, u'initial': [<Storage {u'comment': u'install', u'author': <User: u'/people/admin'>, u'ip': u'127.0.0.1', u'created': datetime.datetime(2013, 3, 20, 10, 27, 4, 640800), u'bot': False, u'changes': u'[]', u'key': u'/type', u'action': u'bulk_update', u'author_id': 24, u'data': u'{}', u'id': 6, u'revision': 1}>], 'created': datetime.datetime(2013, 3, 20, 10, 27, 4, 640800), 'revision': 1, u'recent': []}

versions = h.recent (empty in this case) + h.initial, now the value of versions become h.initial which is a list:
versions = [<Storage {u'comment': u'install', u'author': <User: u'/people/admin'>, u'ip': u'127.0.0.1', u'created': datetime.datetime(2013, 3, 20, 10, 27, 4, 640800), u'bot': False, u'changes': u'[]', u'key': u'/type', u'action': u'bulk_update', u'author_id': 24, u'data': u'{}', u'id': 6, u'revision': 1}>]

latest = versions[0], which gives us a dictionary, which has the value latest.revision=1.

 But these variables are not showing in the '/type/work' so that I can know what is the problem.

 I am really new to all of this stuff, if I have done something wrong, please do tell me, I want to learn and provide value contributions, what I am trying to say is , have I overcomplicated this issue? or have I misunderstood this whole issue?

### So what should I do to make the variables of the '/type/work' page to show the values in the debug=true mode? so that I can understand, what is actually going on.

**Comment by SouthGoingZax** [comment]
It seems like the problem is somewhere in the `page.get_history_preview()` call in lib/history.html.  You can confirm that by deleting lines 3-5 in lib/history.html, and it even works without your %d -> %s change.  A different way to potentially approach the problem (and depending on what you find, maybe a bigger impact change to make) could be to dig into the history (no pun intended) of the history_v2 feature, which has been [set to 'admin'](https://github.com/internetarchive/openlibrary/blob/3580fcf805921c111b2399a068f4d663242c92cc/conf/openlibrary.yml#L64) for many years.  If there are broader benefits to finishing/enabling that feature, it looks like that would resolve this issue too.

**Comment by nk4456542** [comment]
@SouthGoingZax, thank you for your feedback. I will try to solve these based on the above points.

**Comment by mekarpeles** [comment]
Error can be experienced here (incognito):
https://dev.openlibrary.org/type/work?debug=true

Note: the problem doesn't occur when logged in as admin (as noted)

openlibrary/templates/lib/history.html is where the problem is occurring.

https://github.com/internetarchive/openlibrary/blob/master/openlibrary/core/models.py#L68-L105 is the definition for the `get_history_preview` function.

It appears that:

```
$if "superfast" in ctx.features and "history_v2" not in ctx.features:
    $ h = page.get_history_preview()
```
is resulting in a value `h` which is an empty `<type 'instance'>` when it's "empty"... As an admin, it is a `<class 'web.utils.Storage'>`

**Comment by mekarpeles** [comment]
@nk4456542 try out this solution; changing the top of history.html:

```
$def with (page)

$ h = None
$if "superfast" in ctx.features and "history_v2" not in ctx.features:
    $ h = page.get_history_preview()
$if not h:
    $ h = get_history(page)
```

**Comment by nk4456542** [comment]
Thank you for the help @mekarpeles, it is working now, I will submit a PR for this now.

**Comment by SouthGoingZax** [comment]
Just to connect some dots, it looks like #40 is a long-standing bug about history_v2 and why it's not enabled yet.
